core/consensus/src/adapter.rs

```diff
@@ -387,6 +407,189 @@ where
+            return Ok(());
+        };
+
+        if block.header.height != proof.height {
+            return Err(ConsensusError::VerifyBlockProofAndBlockHeightMismatch(
+                block.header.height,
+                proof.height,
+            )
+            .into());
+        }
+
+        let blockhash = Hash::digest(block.clone().encode_fixed()?);
+
+        if blockhash != proof.block_hash {
+            return Err(ConsensusError::VerifyBlockHashMismatch(block.header.height).into());
+        }
+
+        let metadata = self.get_metadata(
+            ctx,
+            block.header.state_root.clone(),
+            block.header.height,
+            block.header.timestamp,
+        )?;
+
+        let mut authority_list = metadata
+            .verifier_list
+            .iter()
+            .map(|v| Node {
+                address:        v.address.as_bytes(),
+                propose_weight: v.propose_weight,
+                vote_weight:    v.vote_weight,
+            })
+            .collect::<Vec<Node>>();
+
+        let signed_voters = extract_voters(&mut authority_list, &proof.bitmap).unwrap();
+
+        let vote_hash = self.crypto.hash(protocol::Bytes::from(rlp::encode(&Vote {
+            height:     proof.height,
+            round:      proof.round,
+            vote_type:  VoteType::Precommit,
+            block_hash: proof.block_hash.as_bytes(),
+        })));
+
+        self.verify_proof_signature(
+            block.header.height,
+            vote_hash,
+            proof.signature.clone(),
+            signed_voters.clone(),
+        )?;
+
+        let weight_map = authority_list
+            .iter()
+            .map(|node| (node.address.clone(), node.vote_weight))
+            .collect::<HashMap<overlord::types::Address, u32>>();
+
+        self.verity_proof_weight(block.header.height, &weight_map, signed_voters)?;
+
+        Ok(())
+    }
+
+    fn verify_proof_signature(
+        &self,
+        block_height: u64,
+        vote_hash: Bytes,
+        aggregated_signature_bytes: Bytes,
+        signed_voters: Vec<Bytes>,
+    ) -> ProtocolResult<()> {
+        // check sig
+        if self
+            .crypto
+            .verify_aggregated_signature(aggregated_signature_bytes, vote_hash, signed_voters)
+            .is_err()
+        {
+            return Err(ConsensusError::VerifyBlockProof(block_height).into());
+        }
+        Ok(())
+    }
+
+    fn verity_proof_weight(
+        &self,
+        block_height: u64,
+        weight_map: &HashMap<Bytes, u32>,
+        signed_voters: Vec<Bytes>,
+    ) -> ProtocolResult<()> {
+        let total_validator_weight: u64 = weight_map.iter().map(|pair| u64::from(*pair.1)).sum();
+
+        let mut accumulator = 0u64;
+        for signed_voter_address in signed_voters {
+            if weight_map.contains_key(signed_voter_address.as_ref()) {
+                accumulator += u64::from(*(weight_map.get(signed_voter_address.as_ref()).unwrap()));
```

Comment:
`weight_map` comes from other nodes, should it panic?

Reply:
it has been verified I remember, but now added an error wrapper